Incrementality for the win!

alxhub opened the pull request.

alxhub committed:
perf(ivy): reuse prior analysis work during incremental builds

Previously, the compiler performed an incremental build by analyzing and
resolving all classes in the program (even unchanged ones) and then using
the dependency graph information to determine which .js files were stale and
needed to be re-emitted. This algorithm produced "correct" rebuilds, but the
cost of re-analyzing the entire program turned out to be higher than
anticipated, especially for component-heavy compilations.

To achieve performant rebuilds, it is necessary to reuse previous analysis
results if possible. Doing this safely requires knowing when prior work is
viable and when it is stale and needs to be re-done.

The new algorithm implemented by this commit is such:

1) Each incremental build starts with knowledge of the last known good
   dependency graph and analysis results from the last successful build,
   plus of course information about the set of files changed.

2) The previous dependency graph's information is used to determine the
   set of source files which have "logically" changed. A source file is
   considered logically changed if it or any of its dependencies have
   physically changed (on disk) since the last successful compilation. Any
   logically unchanged dependencies have their dependency information copied
   over to the new dependency graph.

3) During the `TraitCompiler`'s loop to consider all source files in the
   program, if a source file is logically unchanged then its previous
   analyses are "adopted" (and their 'register' steps are run). If the file
   is logically changed, then it is re-analyzed as usual.

4) Then, incremental build proceeds as before, with the new dependency graph
   being used to determine the set of files which require re-emitting.

This analysis reuse avoids template parsing operations in many circumstances
and significantly reduces the time it takes ngtsc to rebuild a large
application.

Future work will increase performance even more, by tackling a variety of
other opportunities to reuse or avoid work.

JoostK commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -135,256 +135,217 @@ export class IvyCompilation {
+      result = trait.handler.analyze(clazz, trait.detected.metadata);
+    } catch (err) {
+      if (err instanceof FatalDiagnosticError) {
+        trait = trait.toErrored([err.toDiagnostic()]);
```
nit: all `trait` reassignments here are never read from (my IDE shows them grayed out). Since `toErrored` et al mutate the trait and the return value is just to allow proper type inference this isn't a real problem, but on first glance this looks like a bug. I don't think that removing the assignments will really improve the situation here, though.

JoostK commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -135,256 +135,217 @@ export class IvyCompilation {
 
-      // A change to the NgModule file should cause the declaration itself to be invalidated.
-      this.incrementalDriver.trackFileDependency(ngModuleFile, file);
+        if (result.diagnostics !== undefined) {
```
This is undesirable in cases where there's only warning/message diagnostics reported. I don't think we have those currently, but we should probably distinguish here. Can be done in followup work.

alxhub replied:
I don't believe we have any such messages, but yeah, we need to reflect the idea in the trait system that compilation can succeed with warnings.

JoostK commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -135,256 +135,217 @@ export class IvyCompilation {
-        if (match.handler.typeCheck !== undefined && match.analyzed !== null &&
-            match.analyzed.analysis !== undefined) {
-          match.handler.typeCheck(context, node, match.analyzed.analysis);
+  index(ctx: IndexingContext): void {
```
Thanks for moving this down! This has annoyed me for a while :-)

alxhub replied:
👍 you and me both

JoostK commented on packages/compiler-cli/src/ngtsc/transform/src/trait.ts:
```diff
@@ -0,0 +1,149 @@
+  state: TraitState.RESOLVED;
+
+  /**
+   * The result of
```
Unfinished

JoostK commented on packages/compiler-cli/src/ngtsc/transform/src/trait.ts:
```diff
@@ -0,0 +1,149 @@
+ * An implementation of the `Trait` type which transitions safely between the various
+ * `TraitState`s.
+ */
+class TraitImpl<D, A, R> {
```
This is beautiful!

JoostK commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -361,10 +361,13 @@ runInEachFileSystem(os => {
         /**
-         * The following set of tests verify that after Tsickle run we do not have cases which
-         * trigger automatic semicolon insertion, which breaks the code. In order to avoid the
+         * The following set of tests verify that after Tsickle run we do not have cases
```
Indentation has gone weird because of earlier clang-formats. Can you go over this file and revert irrelevant changes?

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -83,31 +84,34 @@ export class DirectiveDecoratorHandler implements
+      inputs: analysis.meta.inputs,
+      outputs: analysis.meta.outputs,
+      queries: analysis.meta.queries.map(query => query.propertyName),
+      isComponent: false, ...extractDirectiveGuards(node, this.reflector),
```
Same here.

alxhub replied:
Also done!

JoostK commented on packages/compiler-cli/src/ngtsc/program.ts:
```diff
@@ -321,6 +316,8 @@ export class NgtscProgram implements api.Program {
       this.perfRecorder.stop(analyzeSpan);
       this.compilation.resolve();
 
+      this.recordNgModuleScopeDependencies();
```
This call is missing from `loadNgStructureAsync` and that is **really** an issue...

alxhub replied:
Fixed!

JoostK commented on packages/compiler-cli/src/ngtsc/shims/src/factory_tracker.ts:
```diff
@@ -0,0 +1,30 @@
+
+import {FactoryGenerator, FactoryInfo} from './factory_generator';
+
+export class FactoryTracker {
```
This one could use a docblock on what information it captures and why that's relevant.

alxhub replied:
Done.

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts:
```diff
@@ -276,6 +277,7 @@ export class NgModuleDecoratorHandler implements
     if (scope !== null) {
       // Using the scope information, extend the injector's imports using the modules that are
       // specified as module exports.
+      // TODO: don't mutate the analysis(!)
```
Agreed!

alxhub replied:
And done!

JoostK commented on packages/compiler-cli/src/ngtsc/file_system/BUILD.bazel:
```diff
@@ -9,7 +9,7 @@ ts_library(
     deps = [
         "//packages:types",
-        "@npm//@types/node",
+      "@npm//@types/node",
```
Indentation is off.

alxhub replied:
Shouldn't have even been changed :)

JoostK commented on packages/compiler-cli/src/ngtsc/incremental/api.ts:
```diff
@@ -0,0 +1,23 @@
+
+export interface IncrementalBuild<W> {
+  /**
+   *
```
Pending a comment

JoostK commented on packages/compiler-cli/src/ngtsc/incremental/src/README.md:
```diff
@@ -92,3 +92,13 @@ For example, today a component's `NgModule` and all of the other components whic
+
+* write a commit description
+* update this README to describe the new mechanisms, as now it's horribly out of date
+* share a module resolution cache between ngtsc & the cycle detector
```
I have this in a separate PR!

alxhub replied:
🚀

JoostK commented on packages/compiler-cli/src/ngtsc/incremental/src/dependency_tracking.ts:
```diff
@@ -0,0 +1,117 @@
+    }
+  }
+
+  getDependencies(sf: T): ReadonlySet<string> { return this.metadataFor(sf).dependsOn; }
```
This isn't currently used.

alxhub replied:
Gone!

JoostK commented on packages/compiler-cli/src/ngtsc/incremental/src/dependency_tracking.ts:
```diff
@@ -0,0 +1,117 @@
+
+export class FileDependencyGraph<T extends{fileName: string} = ts.SourceFile> implements
+    DependencyTracker<T> {
+  private metadata = new Map<T, FileMetadata>();
```
I feel that the term `metadata` is no longer correct in this context. It's now the concrete dependency data, crucial to a dependency graph. How about something like `nodes` and rename `FileMetadata` into `DependencyNode` or something?

alxhub replied:
+1. I went with `FileNode` since it seems to fit.

JoostK commented on packages/compiler-cli/src/ngtsc/incremental/src/dependency_tracking.ts:
```diff
@@ -0,0 +1,117 @@
+    return isChanged(sf, this.metadataFor(sf), changedTsPaths, EMPTY_SET, changedResources);
+  }
+
+  updateWithPhysicalChanges(
```
This is so nicely done! 👏

alxhub replied:
Thank you! I added some comments describing the behavior, since it's a little convoluted (the method both mutates the current state, and returns some information).

JoostK commented on packages/compiler-cli/src/ngtsc/incremental/src/dependency_tracking.ts:
```diff
@@ -0,0 +1,117 @@
+    return logicallyChanged;
+  }
+
+  clone(): FileDependencyGraph<T> {
```
I don't see this being used anymore.

alxhub replied:
You are absolutely right, it's not.

JoostK commented on packages/compiler-cli/src/ngtsc/incremental/src/state.ts:
```diff
@@ -8,26 +8,25 @@
 
-import {DependencyTracker} from '../../partial_evaluator';
-import {ResourceDependencyRecorder} from '../../util/src/resource_recorder';
+import {ClassRecord, TraitCompiler} from '../../transform/src/compilation';
```
These should be exported from the entry-point of the `transform` target.

alxhub replied:
Good catch!

JoostK commented on packages/compiler-cli/src/ngtsc/incremental/src/state.ts:
```diff
@@ -155,66 +173,34 @@ export class IncrementalDriver implements DependencyTracker, ResourceDependencyR
       pendingEmit,
+
+      // Since this compilation was successfully analyzed, update the "last good" artifacts to those
+      // to the current ones.
```
This sentence seems incorrect.

alxhub replied:
Fixed!

JoostK commented on packages/compiler-cli/src/ngtsc/program.ts:
```diff
@@ -640,44 +641,45 @@ export class NgtscProgram implements api.Program {
+        // remote scoping feature which is activated in the event of potential import cycles. Thus,
+        // the module depends not only on the transitive dependencies of the component, but on its
+        // resources as well.
+        depGraph.addTransitiveResources(ngModuleFile, file);
```
Nice catch! This need a test, though.

alxhub replied:
Agreed. I'm gonna do this in a follow-up though (when we do the proper module graph logic).

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/src/component.ts:
```diff
@@ -13,15 +13,16 @@ import {CycleAnalyzer} from '../../cycles';
 import {EnumValue, PartialEvaluator} from '../../partial_evaluator';
 import {ClassDeclaration, Decorator, ReflectionHost, reflectObjectLiteral} from '../../reflection';
 import {ComponentScopeReader, LocalModuleScopeRegistry} from '../../scope';
+import {FactoryTracker} from '../../shims';
```
This has gone unused.

alxhub commented on packages/compiler-cli/ngcc/src/analysis/decoration_analyzer.ts:
```diff
@@ -212,15 +218,17 @@ export class DecorationAnalyzer {
           if (diagnostics !== undefined) {
             diagnostics.forEach(error => this.diagnosticHandler(error));
           }
+          match.resolution = data !;
```
It's needed because for some reason, `undefined` is assignable to `unknown`, but not to `Readonly<unknown>`. I'm now using an explicit cast to that instead.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/component.ts:
```diff
@@ -70,6 +75,7 @@ export class ComponentDecoratorHandler implements
   private preanalyzeTemplateCache = new Map<ts.Declaration, PreanalyzedTemplate>();
 
   readonly precedence = HandlerPrecedence.PRIMARY;
+  readonly name = ComponentDecoratorHandler.name;
```
Nope!

`readonly name` is not static. This is using the name of the constructor function (class) `ComponentDecoratorHandler`, not the instance field `name`.

JoostK replied:
Ah, right. This is a bug in my IDE, then, as it clicks through to the declaration of the field itself (TypeScript itself does the right thing, so it's an IntelliJ bug)

alxhub committed:
refactor(ivy): formalize the compilation process for matched handlers

Prior to this commit, the `IvyCompilation` tracked the state of each matched
`DecoratorHandler` on each class in the `ts.Program`, and how they
progressed through the compilation process. This tracking was originally
simple, but had grown more complicated as the compiler evolved. The state of
each specific "target" of compilation was determined by the nullability of
a number of fields on the object which tracked it.

This commit formalizes the process of compilation of each matched handler
into a new "trait" concept. A trait is some aspect of a class which gets
created when a `DecoratorHandler` matches the class. It represents an Ivy
aspect that needs to go through the compilation process.

Traits begin in a "pending" state and undergo transitions as various steps
of compilation take place. The `IvyCompilation` class is renamed to the
`TraitCompiler`, which manages the state of all of the traits in the active
program.

Making the trait concept explicit will support future work to incrementalize
the expensive analysis process of compilation.

alxhub committed:
refactor(ivy): move analysis side effects into a register phase

Previously 'analyze' in the various `DecoratorHandler`s not only extracts
information from the decorators on the classes being analyzed, but also has
several side effects within the compiler:

* it can register metadata about the types involved in global metadata
  trackers.
* it can register information about which .ngfactory symbols are actually
  needed.

In this commit, these side-effects are moved into a new 'register' phase,
which runs after the 'analyze' step. Currently this is a no-op refactoring
as 'register' is always called directly after 'analyze'. In the future this
opens the door for re-use of prior analysis work (with only 'register' being
called, to apply the above side effects).

Also as part of this refactoring, the reification of NgModule scope
information into the incremental dependency graph is moved to the
`NgtscProgram` instead of the `TraitCompiler` (which now only manages trait
compilation and does not have other side effects).

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/component.ts:
```diff
@@ -294,27 +288,40 @@ export class ComponentDecoratorHandler implements
+      inputs: analysis.meta.inputs,
+      outputs: analysis.meta.outputs,
+      queries: analysis.meta.queries.map(query => query.propertyName),
+      isComponent: true, ...extractDirectiveGuards(node, this.reflector),
```
Agreed, and done.

alxhub commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -19,111 +19,111 @@ import {LocalModuleScopeRegistry} from '../../scope';
 /**
- * Record of an adapter which decided to emit a static field, and the analysis it performed to
- * prepare for that operation.
+ * A t
```
Finished!

JoostK commented on packages/compiler-cli/src/ngtsc/incremental/src/README.md:
```diff
@@ -16,6 +45,14 @@ The second condition is challenging to prove, as Angular allows statically evalu
+
+To reuse previous analyses, ngtsc uses the _prior_ compilation's dependency graph, plus the information about which files have changed, to determine whether it's safe to reuse the prior compilation's work.
+
+To skip emit, ngtsc uses the _current_ compilation's dependency graph to determine whether (given the information about which files have changed), a particular file has become stale on disk and needs to be re-emitted.
```
The part within parens here reads a bit awkward.

alxhub replied:
Reworded!

JoostK commented on packages/compiler-cli/src/ngtsc/incremental/src/dependency_tracking.ts:
```diff
@@ -0,0 +1,145 @@
+   * where:
+   *
+   * G(n) = the dependency graph of build `n`
+   * L(n) = the logically changed files from build n - 1 to build n.
```
Referring to build `n - 1` as representing the reference build for logical changes is not strictly true, as any changes during failed builds will accumulate.

JoostK commented on packages/compiler-cli/src/ngtsc/incremental/src/dependency_tracking.ts:
```diff
@@ -0,0 +1,145 @@
+      const node = previous.nodeFor(sf);
+      if (isLogicallyChanged(sf, node, changedTsPaths, deletedTsPaths, changedResources)) {
+        logicallyChanged.add(sf.fileName);
+      } else if (!deletedTsPaths.has(sf.fileName)) {
```
This condition is always true, given the first condition in `isLogicallyChanged`.

alxhub replied:
Well spotted!

This is a holdover from when the dep graph was first copied and then updated with changes. At some point I rewrote the algorithm to start with an empty graph and populate it while accounting for changes, at which point deletion is no longer needed.

JoostK commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -19,111 +19,138 @@ import {LocalModuleScopeRegistry} from '../../scope';
  *
- * The compilation is stateful - source files are analyzed and records of the operations that need
- * to be performed during the transform/emit process are maintained internally.
+ * The `TraitCompiler` is responsible for processing all classes in the program and
```
Unfinished sentence

alxhub replied:
Finished!

JoostK reviewed (approved):
Typo in commit message of ad28e7a79b0b4d65357714e5b48c77e3d646c571:

but had grown more _compilated_ -> but had grown more **complicated**

JoostK commented:
One more thing: can you extract this logic  out into a method so that it can be shared between `loadNgStructureAsync` and `ensureAnalyzed`:

```ts
    this.compilation.resolve();

    this.recordNgModuleScopeDependencies();

    // At this point, analysis is complete and the compiler can now calculate which files need to be
    // emitted, so do that.
    this.incrementalDriver.recordSuccessfulAnalysis(this.compilation);
```

alxhub commented:
> compilated

Technically, we do work on the compiler, so the code does get more _compilated_ :-P 

Thanks, fixed!

alxhub commented:
[Presubmit](http://test/OCL:285217442:BASE:285223078:1576176387120:87ad3dcb)
[Ivy Presubmit](http://test/OCL:285217442:BASE:285223078:1576176387120:af9a71b7)

alxhub commented:
merge-assistance: I am the owner of the code in question.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/component.ts:
```diff
@@ -294,12 +311,9 @@ export class ComponentDecoratorHandler implements
 
           // These will be replaced during the compilation step, after all `NgModule`s have been
           // analyzed and the full compilation scope for the component can be realized.
```
Should this comment be removed?

JoostK replied:
Agreed, it should have been removed. Nicely spotted!
